packages/opentelemetry-core/src/context/propagation/BinaryTraceContext.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BinaryFormat, SpanContext } from '@opentelemetry/types';
+
+const VERSION_ID = 0;
+const TRACE_ID_FIELD_ID = 0;
+const SPAN_ID_FIELD_ID = 1;
+const TRACE_OPTION_FIELD_ID = 2;
+
+// Sizes are number of bytes.
+const ID_SIZE = 1;
+const TRACE_ID_SIZE = 16;
+const SPAN_ID_SIZE = 8;
+const TRACE_OPTION_SIZE = 1;
+
+const VERSION_ID_OFFSET = 0;
+const TRACE_ID_FIELD_ID_OFFSET = VERSION_ID_OFFSET + ID_SIZE;
+const TRACE_ID_OFFSET = TRACE_ID_FIELD_ID_OFFSET + ID_SIZE;
+const SPAN_ID_FIELD_ID_OFFSET = TRACE_ID_OFFSET + TRACE_ID_SIZE;
+const SPAN_ID_OFFSET = SPAN_ID_FIELD_ID_OFFSET + ID_SIZE;
+const TRACE_OPTION_FIELD_ID_OFFSET = SPAN_ID_OFFSET + SPAN_ID_SIZE;
+const TRACE_OPTIONS_OFFSET = TRACE_OPTION_FIELD_ID_OFFSET + ID_SIZE;
+
+const FORMAT_LENGTH =
+  4 * ID_SIZE + TRACE_ID_SIZE + SPAN_ID_SIZE + TRACE_OPTION_SIZE;
+
+export class BinaryTraceContext implements BinaryFormat {
+  toBytes(spanContext: SpanContext): Buffer {
```

Comment:
Actually, there could have an interface that is compatible with other propagators with something like this:

```js
inject(spanContext: SpanContext, format: string, carrier: Injector) {
```

And then the injector would be a function/class passed by the user that would take the buffer as input. Then they can control how to add the buffer created by OpenTelemetry in their own existing buffer, while keeping the API compatible.

Reply:
This is based on specs here https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/propagators-api.md#propagators-api. Even I was confused about these two interfaces (`BinaryFormat` and `HTTPTextFormat`), but decided to have something ready for discussion.

@bogdandrutu Could you please address these questions?

Reply:
This could be done similar to OpenTracing, but with an `ArrayBuffer` to support the browser too:

```ts
interface BinaryCarrier {
  buffer: ArrayBuffer
}

class BinaryPropagator implements Propagator {
  inject(spanContext: SpanContext, format: string, carrier: BinaryCarrier): this { ... }
  extract(format: string, carrier: BinaryCarrier): SpanContext | null { ... }
}
```

Reply:
Sure makes sense to me, will wait for https://github.com/open-telemetry/opentelemetry-specification/issues/174 to get accepted. Mean time, Could you please review again.

Reply:
If I recall correctly, the reason in OpenCensus that this was a `Buffer` was to support gRPC, which isn't relevant in the browser case (there is gRPC web but its transport is actually different from native gRPC and it uses a proxy that speaks HTTP actually)

Reply:
A `Buffer` can be initialized from an `ArrayBuffer` in Node and same for a `TypedArray` in the browser. The idea then would be to use an `ArrayBuffer` which is compatible with both platforms, and assign it to an object instead of returning it to make it compatible with `inject` and `extract`.

I think we can move forward with these APIs in the meantime while https://github.com/open-telemetry/opentelemetry-specification/issues/174 and https://github.com/open-telemetry/opentelemetry-specification/issues/165 are being discussed

Reply:
@mayurkale22 could you make this an `ArrayBuffer` per our conversation in the SIG today?

It is browser compatible and is efficiently convertable to a Node `Buffer` via `Buffer.from`

Reply:
Done, PTAL